salt/cache/etcd_cache.py

```diff
@@ -0,0 +1,216 @@
+with a read-write access to the key-value store.
+
+The related documentation can be found in the `Etcd documentation`_.
+
+To enable this cache plugin, the master will need the python client for
+Etcd installed. This can be easily installed with pip:
+
+.. code-block: bash
+
+    pip install python-etcd
+
+Optionally, depending on the Etcd agent configuration, the following values
+could be set in the master config. These are the defaults:
+
+.. code-block:: yaml
+
+    etcd.host: 127.0.0.1
+    etcd.port: 2379
+    etcd.protocol: http
+    etcd.allow_reconnect: True
+    etcd.allow_redirect: False
+    etcd.srv_domain: None
+    etcd.read_timeout: 60
+    etcd.username: None
+    etcd.password: None
+    etcd.cert: None
+    etcd.ca_cert: None
+
+Related docs could be found in the `python-etcd documentation`_.
+
+To use the etcd as a minion data cache backend, set the master ``cache`` config
+value to ``etcd``:
+
+.. code-block:: yaml
+
+    cache: etcd
+
+
+.. _`Etcd documentation`: https://github.com/coreos/etcd
+.. _`python-etcd documentation`: http://python-etcd.readthedocs.io/en/latest/
+
+'''
+from __future__ import absolute_import
+import logging
+try:
+    import etcd
+    HAS_ETCD = True
+except ImportError:
+    HAS_ETCD = False
+
+from salt.exceptions import SaltCacheError
+
+_DEFAULT_PATH_PREFIX = "/salt_cache"
+
+if HAS_ETCD:
+    # The client logging tries to decode('ascii') binary data
+    # and is too verbose
+    etcd._log.setLevel(logging.INFO)  # pylint: disable=W0212
+
+log = logging.getLogger(__name__)
+client = None
+path_prefix = None
+
+# Module properties
+
+__virtualname__ = 'etcd'
+__func_alias__ = {'ls': 'list'}
+
+
+def __virtual__():
+    '''
+    Confirm that python-etcd package is installed. Setup client and init
+    datastore.
+    '''
+    if not HAS_ETCD:
+        return (False, "Please install python-etcd package to use etcd data "
+                "cache driver")
+
+    etcd_kwargs = {
+            'host': __opts__.get('etcd.host', '127.0.0.1'),
+            'port': __opts__.get('etcd.port', 2379),
+            'protocol': __opts__.get('etcd.protocol', 'http'),
+            'allow_reconnect': __opts__.get('etcd.allow_reconnect', True),
+            'allow_redirect': __opts__.get('etcd.allow_redirect', False),
+            'srv_domain': __opts__.get('etcd.srv_domain', None),
+            'read_timeout': __opts__.get('etcd.read_timeout', 60),
+            'username': __opts__.get('etcd.username', None),
+            'password': __opts__.get('etcd.password', None),
+            'cert': __opts__.get('etcd.cert', None),
+            'ca_cert': __opts__.get('etcd.ca_cert', None),
+    }
+
+    global client, path_prefix
+    path_prefix = __opts__.get('etcd.path_prefix', _DEFAULT_PATH_PREFIX)
+    if path_prefix != "":
+        path_prefix = '/{0}'.format(path_prefix.strip('/'))
+    log.info("etcd: Setting up client with params: %r", etcd_kwargs)
+    client = etcd.Client(**etcd_kwargs)
```

Comment:
Same issue here as in the other PR. We should not be doing this sort of initialization in `__virtual__` IMHO.